File: backend/src/main/java/com/noisevisionsoftware/vitema/service/product/ProductService.java
package com.noisevisionsoftware.vitema.service.product;

import com.noisevisionsoftware.vitema.dto.product.IngredientDTO;
import com.noisevisionsoftware.vitema.dto.request.product.ProductRequest;
import com.noisevisionsoftware.vitema.dto.response.product.ProductResponse;
import com.noisevisionsoftware.vitema.model.product.Product;
import com.noisevisionsoftware.vitema.model.product.ProductType;
import com.noisevisionsoftware.vitema.model.recipe.NutritionalValues;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Service for managing products.
 * Refactored to use PostgreSQL (via ProductDatabaseService) instead of Firestore.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class ProductService {

    private final ProductDatabaseService productDatabaseService;

    /**
     * Returns GLOBAL products + CUSTOM products belonging to the specific trainer.
     */
    public List<IngredientDTO> searchProducts(String query, String trainerId) {
        if (query == null || query.trim().isEmpty()) {
            return List.of();
        }

        // Using database service to search by name
        List<ProductResponse> products = productDatabaseService.searchByName(query, trainerId);

        return products.stream()
                .map(this::mapResponseToIngredientDTO)
                .collect(Collectors.toList());
    }

    /**
     * Automatically determines type (GLOBAL/CUSTOM) based on who is asking.
     */
    public Product createProduct(Product product, String trainerId) {
        if (product.getName() == null || product.getName().trim().isEmpty()) {
            throw new IllegalArgumentException("Product name is required");
        }

        ProductType type = ProductType.GLOBAL;
        String authorId = null;

        if (trainerId != null && !trainerId.isEmpty()) {
            type = ProductType.CUSTOM;
            authorId = trainerId;
        }

        ProductRequest request = getProductRequest(product, type);

        ProductResponse savedResponse = productDatabaseService.create(request, authorId, type);

        return mapResponseToProduct(savedResponse);
    }

    private static ProductRequest getProductRequest(Product product, ProductType type) {
        ProductRequest request = new ProductRequest();
        request.setName(product.getName());
        request.setCategory(product.getCategoryId());
        request.setUnit(product.getDefaultUnit() != null ? product.getDefaultUnit() : "g");

        // Default verification logic: Global = Verified, Custom = Unverified
        request.setVerified(type == ProductType.GLOBAL);

        if (product.getNutritionalValues() != null) {
            request.setKcal(product.getNutritionalValues().getCalories());
            request.setProtein(product.getNutritionalValues().getProtein());
            request.setFat(product.getNutritionalValues().getFat());
            request.setCarbs(product.getNutritionalValues().getCarbs());
        }
        return request;
    }

    /**
     * Retrieves a product by ID (converts String ID to Long).
     */
    public Optional<Product> getProductById(String id) {
        try {
            return productDatabaseService.findById(Long.parseLong(id))
                    .map(this::mapResponseToProduct);
        } catch (NumberFormatException e) {
            return Optional.empty();
        }
    }

    /**
     * ENFORCES ownership security. A trainer cannot delete a Global product
     * or another trainer's product.
     */
    public void deleteProduct(String id, String trainerId) {
        try {
            Long dbId = Long.parseLong(id);

            Optional<ProductResponse> existingOpt = productDatabaseService.findById(dbId);

            if (existingOpt.isEmpty()) {
                throw new IllegalArgumentException("Product not found");
            }

            ProductResponse existing = existingOpt.get();

            boolean isGlobal = existing.getType() == ProductType.GLOBAL;
            boolean isOwner = trainerId != null && trainerId.equals(existing.getAuthorId());

            if (isGlobal) {
                throw new IllegalArgumentException("Cannot delete GLOBAL products");
            }

            if (!isOwner) {
                throw new IllegalArgumentException("You can only delete your own custom products");
            }

            productDatabaseService.delete(dbId);

        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Invalid product ID: " + id);
        }
    }

    private IngredientDTO mapResponseToIngredientDTO(ProductResponse response) {
        return IngredientDTO.builder()
                .id(String.valueOf(response.getId()))
                .name(response.getName())
                .defaultUnit(response.getUnit())
                .nutritionalValues(NutritionalValues.builder()
                        .calories(response.getKcal())
                        .protein(response.getProtein())
                        .fat(response.getFat())
                        .carbs(response.getCarbs())
                        .build())
                .categoryId(response.getCategory())
                .type(response.getType() != null ? response.getType().name() : ProductType.GLOBAL.name())
                .build();
    }

    private Product mapResponseToProduct(ProductResponse response) {
        return Product.builder()
                .id(String.valueOf(response.getId()))
                .name(response.getName())
                .searchName(response.getName().toLowerCase())
                .defaultUnit(response.getUnit())
                .nutritionalValues(NutritionalValues.builder()
                        .calories(response.getKcal())
                        .protein(response.getProtein())
                        .fat(response.getFat())
                        .carbs(response.getCarbs())
                        .build())
                .categoryId(response.getCategory())
                .type(response.getType() != null ? response.getType() : ProductType.GLOBAL)
                .authorId(response.getAuthorId())
                .build();
    }
}

File: frontend/src/services/product/ProductService.ts
import api from "../../config/axios";
import {Product} from "../../types/product";

export interface ProductSearchParams {
    query: string;
    trainerId?: string;
    limit?: number;
}

export interface CreateProductRequest {
    name: string;
    defaultUnit: string;
    nutritionalValues: {
        calories: number;
        protein: number;
        fat: number;
        carbs: number;
    };
    categoryId?: string;
}

export class ProductService {
    private static readonly BASE_URL = '/products';

    /**
     * Search products from the local Firestore database
     * @param params - Search parameters including query and optional trainerId
     * @returns List of products matching the search criteria
     */
    static async searchProducts(params: ProductSearchParams): Promise<Product[]> {
        try {
            const response = await api.get(`${this.BASE_URL}/search`, {
                params: {
                    query: params.query,
                    trainerId: params.trainerId,
                    limit: params.limit || 20
                }
            });
            return response.data;
        } catch (error) {
            console.error('Error searching products:', error);
            throw error;
        }
    }

    /**
     * Get a single product by ID
     * @param id - Product ID
     */
    static async getProductById(id: string): Promise<Product> {
        const response = await api.get(`${this.BASE_URL}/${id}`);
        return response.data;
    }

    /**
     * Create a custom product (for trainers)
     * @param product - Product data
     * @param trainerId - Trainer ID (will be set as author)
     */
    static async createCustomProduct(
        product: CreateProductRequest,
        trainerId: string
    ): Promise<Product> {
        const response = await api.post(`${this.BASE_URL}`, product, {
            params: {trainerId}
        });
        return response.data;
    }

    /**
     * Delete a custom product
     * @param id - Product ID
     * @param trainerId - Trainer ID (must be the author)
     */
    static async deleteProduct(id: string, trainerId: string): Promise<void> {
        await api.delete(`${this.BASE_URL}/${id}`, {
            params: {trainerId}
        });
    }

    /**
     * Seed initial products (admin/dev only)
     * Note: Products are automatically seeded on first application startup
     */
    static async seedBasicProducts(): Promise<string> {
        const response = await api.post(`${this.BASE_URL}/seed/basic`);
        return response.data;
    }
}
